File: systemc/TLM_router.cpp
#include "tlm_router/initiator.h"
#include "tlm_router/target.h"
#include "tlm_router/router.h"

SC_MODULE(Top)
{
  Initiator* initiator;
  Router<4>* router;
  Memory*    memory[4];

  SC_CTOR(Top)
  {
    // Instantiate components
    initiator = new Initiator("initiator");
    router    = new Router<4>("router");
    for (int i = 0; i < 4; i++)
    {
      char txt[20];
      sprintf(txt, "memory_%d", i);
      memory[i]   = new Memory(txt);
    }

    // Bind sockets
    initiator->socket.bind( router->target_socket );
    for (int i = 0; i < 4; i++)
      router->initiator_socket[i]->bind( memory[i]->socket );
  }
};

int sc_main(int argc, char* argv[])
{
  Top top("top");
  sc_start();
  return 0;
}

File: simpy/TLM_router.py
from simpy import Environment
from tlmrouter.initiator import *
from tlmrouter.target import *
from tlmrouter.router import *
from utils import *

class Top(Module):
    def __init__(self, env: Environment, name: str):
        super().__init__(env, name)
        self.initiator = Initiator(self.env, "init")
        self.memory = [None] * 4
        self.router = Router(self.env, "router", 4)
        for i in range(4):
            s = f"mem_{i}"
            self.memory[i] = Memory(self.env, s)
        self.initiator.socket.bind(self.router.target_socket)
        for i in range(4):
            self.router.initiator_socket[i].bind(self.memory[i].socket)


if __name__ == "__main__":
    env = simpy.Environment()
    top = Top(env, "top")
    env.run()
